File: examples/localizer.lua
config = {physics = {gravity = 9.8, friction = 0.98}}
posX = {0, 10}
posY = {0, 5}
velX = {1, 2}
velY = {0, 0}
count = 2
dt = 0.5
local ____config_physics_friction = config.physics.friction
for i = 1, count do
    local ____velX = velX[i]
    local ____velY = velY[i]
    ____velX = ____velX * ____config_physics_friction
    ____velY = ____velY + config.physics.gravity * dt
    ____velY = ____velY * ____config_physics_friction
    posX[i] = posX[i] + ____velX * dt
    posY[i] = posY[i] + ____velY * dt
    velX[i] = ____velX
    velY[i] = ____velY
end
terminalSpeed = config.physics.gravity / config.physics.friction
drag = {1, 0.9}
function applyDrag(i)
    drag[i + 1] = drag[i + 1] * 0.5
end
for i = 0, count - 1 do
    drag[i + 1] = drag[i + 1] * config.physics.friction
    applyDrag(i)
end
print(posX[1], posY[1])
print(terminalSpeed)
print(drag[1], drag[2])

File: examples/localizer.ts
const config = {
  physics: {
    gravity: 9.8,
    friction: 0.98,
  },
};
const posX = [0, 10];
const posY = [0, 5];
const velX = [1, 2];
const velY = [0, 0];
const count = 2;
const dt = 0.5;

// config.physics.friction accessed 2x in loop → hoisted to a local above
// the loop. velX[i] and velY[i] read and written multiple times → cached
// in locals within the loop body, written back at end of each iteration.
for (const i of $range(0, count - 1)) {
  velX[i] = velX[i] * config.physics.friction;
  velY[i] = velY[i] + config.physics.gravity * dt;
  velY[i] = velY[i] * config.physics.friction;
  posX[i] = posX[i] + velX[i] * dt;
  posY[i] = posY[i] + velY[i] * dt;
}

// config.physics.gravity accessed only 1x per scope → not hoisted
const terminalSpeed = config.physics.gravity / config.physics.friction;

// --- Limitations ---

// Not localized: applyDrag() mutates the drag array between reads — caching
// drag[i] in a local would miss the write from the function call
const drag = [1.0, 0.9];
function applyDrag(i: number): void {
  drag[i] = drag[i] * 0.5;
}
for (const i of $range(0, count - 1)) {
  drag[i] = drag[i] * config.physics.friction;
  applyDrag(i);
}

print(posX[0], posY[0]); // 0.49  2.401
print(terminalSpeed); // 10
print(drag[0], drag[1]); // 0.49  0.441
